src/QirRuntime/lib/QIR/callables.cpp

```diff
@@ -191,37 +56,207 @@ extern "C"
     }
 
-    void quantum__rt__callable_invoke(QirCallable* clb, QirTupleHeader* args, QirTupleHeader* result)
+    void quantum__rt__callable_invoke(QirCallable* callable, QirTupleHeader* args, QirTupleHeader* result)
     {
-        clb->Invoke(args, result);
+        assert(callable != nullptr);
+        callable->Invoke(args, result);
     }
 
     QirCallable* quantum__rt__callable_copy(QirCallable* other)
     {
-        return new QirCallable(other);
+        if (other == nullptr)
+        {
+            return nullptr;
+        }
+        return new QirCallable(*other);
     }
 
-    void quantum__rt__callable_make_adjoint(QirCallable* clb)
+    void quantum__rt__callable_make_adjoint(QirCallable* callable)
     {
-        clb->ApplyFunctor(QirCallable::Adjoint);
+        assert(callable != nullptr);
+        callable->ApplyFunctor(QirCallable::Adjoint);
     }
 
-    void quantum__rt__callable_make_controlled(QirCallable* clb)
+    void quantum__rt__callable_make_controlled(QirCallable* callable)
     {
-        clb->ApplyFunctor(QirCallable::Controlled);
+        assert(callable != nullptr);
+        callable->ApplyFunctor(QirCallable::Controlled);
+    }
+}
+
+/*==============================================================================
+    Implementation of QirCallable
+==============================================================================*/
+QirCallable::~QirCallable()
+{
+    assert(refCount == 0);
+}
+
+QirCallable::QirCallable(const t_CallableEntry* ftEntries, QirTupleHeader* capture)
+    : refCount(1)
+    , capture(capture)
+    , appliedFunctor(0)
+    , controlledDepth(0)
+{
+    memcpy(this->functionTable, ftEntries, QirCallable::TableSize * sizeof(void*));
+    assert(this->functionTable[0] != nullptr); // base must be always defined
+}
+
+QirCallable::QirCallable(const QirCallable& other)
+    : refCount(1)
+    , capture(other.capture)
+    , appliedFunctor(other.appliedFunctor)
+    , controlledDepth(other.controlledDepth)
+{
+    memcpy(this->functionTable, other.functionTable, QirCallable::TableSize * sizeof(void*));
+}
+
+long QirCallable::AddRef()
+{
+    long rc = ++this->refCount;
+    assert(rc != 1); // not allowed to resurrect!
+    return rc;
+}
+
+long QirCallable::Release()
+{
+    assert(this->refCount > 0);
+
+    long rc = --this->refCount;
+    if (rc == 0)
+    {
+        delete this;
+    }
+    return rc;
+}
+
+// `src` tuple header must point to a tuple with the following structure (example for depth = 2):
+// { %TupleHeader, %Array*, { %TupleHeader, %Array*, { %TupleHeader, i64, %Qubit* }* }* }
+// or
+// { %TupleHeader, %Array*, { %TupleHeader, %Array*, %Qubit* }* }
+// The function will create a new tuple, where the array contains elements of all nested arrays, respectively:
+// { %TupleHeader, %Array*, { %TupleHeader, i64, %Qubit* }* }
+// { %TupleHeader, %Array*, %Qubit* }
+// The caller is responsible for releasing both the new array and the tuple.

```

Comment:
Suggest adding explicit mention that caller shouldn't free last element. But be careful with wording.

Reply:
The caller cannot release the last element because they don't know what that is.